[inductor] add KernelTemplateParams

coconutruben opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #156151
* #156273
* #156275
* #156150
* __->__ #155775
* #155774
* #155773

Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run @mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

coconutruben committed:
[inductor] add KernelTemplateParams

Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run @mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] add KernelTemplateParams"


Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

NikhilAPatel reviewed (approved):
LGTM

jansel reviewed (changes requested):
Failing tests?

coconutruben commented:
> Failing tests?

- yeah I think some of the param changes the order of the kwargs when we print, so I'll change that out to match again
- some compile time changes (better?) so we need to regenerate

but this one @jansel if you have pointers, I've come across this before in development but always side stepped it somehow, but fundamentally I don't know what causes this behavior

```
PRINTING LOG FILE of inductor/test_cooperative_reductions 1/1 (test/test-reports/inductor.test_cooperative_reductions_1.1_7284db248664bb27_.log)
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/inductor/test_cooperative_reductions.py", line 10, in <module>
    from torch._inductor.choices import InductorChoices
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/choices.py", line 15, in <module>
    from .template_heuristics import (
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/template_heuristics.py", line 12, in <module>
    from .kernel.kernel_params import (
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/kernel/__init__.py", line 1, in <module>
    from . import mm, mm_common, mm_plus_mm
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/kernel/mm.py", line 17, in <module>
    from torch._inductor.codegen.cpp_gemm_template import CppGemmTemplate
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/codegen/cpp_gemm_template.py", line 14, in <module>
    from .. import config, ir, lowering as L
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/lowering.py", line 7109, in <module>
    from . import quantized_lowerings
  File "/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/quantized_lowerings.py", line 8, in <module>
    from .codegen.cpp_gemm_template import CppGemmTemplate, CppWoqInt4GemmTemplate
ImportError: cannot import name 'CppGemmTemplate' from partially initialized module 'torch._inductor.codegen.cpp_gemm_template' (most likely due to a circular import) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/_inductor/codegen/cpp_gemm_template.py)
Got exit code 1
```

the `ImportError: cannot import name 'CppGemmTemplate' from partially initialized module` - is it just circular imports and I need to just sort through it, or something else on the pt side that can cause this

masnesral commented:
> is it just circular imports and I need to just sort through it, or something else on the pt side that can cause this

@coconutruben this just looks like a circular import to me; i.e.,  cpp_gemm_template.py trying to import itself (via lowerings.py, quantized_lowerings.py, etc.)

coconutruben committed:
Update on "[inductor] add KernelTemplateParams"


Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] add KernelTemplateParams"


Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben commented:
tl;dr import cycle related to lowering and cpp_template, sidestepping for now while we refactor and addressing when cpp_template is part of the refactor

longer: 

I think there is a real import cycle that gets triggered through cpp_template that shows up when we try to import the kernel module too early potentially. It's not clear to me how to untangle this right now (lowering -> quantized_lowering -> cpp_template -> lowering) etc but in general it seems when we import kernel, so mm, mm_options etc too early we get into a mess because of cpp_template.
I'm sidestepping this for now, and in subsequent PRs, as we're pushing all these things inside template_heuristics and triton-templates. When we're at the phase of untangling the cpp template, it'll become clearer how to untangle this without running into the import cycle

coconutruben committed:
Update on "[inductor] add KernelTemplateParams"


Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben committed:
Update on "[inductor] add KernelTemplateParams"


Summary:
\# Why

This abstraction will gather all the parameters we need to pass to a KernelTemplate to generate a kernel (and ChoiceCaller). This allows us to serialize/deserialize these parameters easily when doing a lookup table, and pass them to a performance model when doing prediction

\# What

- add new classes to capture exactly the kwargs being passed to different templates
- use those new classes through a new interface (*mm_params) in addmm and mm
- as the rest of the stack: focus is mm and addmm initially, before porting to others. This is why we're keeping around the old interface (*mm_configs) until everything uses the new one

\# Context

- this is part of the effort to prepare everything for a lookup table and a performance prediction model inside InductorChoices
- the approach is to focus on Triton, mm and addmm first, before fanning out to others

Test Plan:
```
buck2 run mode/opt scripts/coconutruben/torchmm:experiment 2>&1
```
```
AUTOTUNE mm(4096x3008, 3008x2048)
strides: [3008, 1], [2048, 1]
dtypes: torch.float16, torch.float16
  mm 0.0729 ms 100.0%
  triton_mm_persistent_tma_25 0.0759 ms 96.1% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=256, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_19 0.0762 ms 95.8% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=256, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_17 0.0787 ms 92.7% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_26 0.0794 ms 91.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_24 0.0802 ms 91.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_23 0.0844 ms 86.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=True, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=4, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_22 0.0874 ms 83.5% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=4, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_persistent_tma_21 0.0879 ms 83.0% ACC_TYPE='tl.float32', ALLOW_TF32=False, A_ROW_MAJOR=True, BLOCK_K=128, BLOCK_M=128, BLOCK_N=128, B_ROW_MAJOR=True, EVEN_K=False, GROUP_M=8, NUM_SMS=132, TMA_SIZE=128, USE_FAST_ACCUM=False, num_stages=3, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
  triton_mm_18 0.0890 ms 81.9% ACC_TYPE='tl.float32', ALLOW_TF32=False, BLOCK_K=64, BLOCK_M=128, BLOCK_N=128, EVEN_K=True, GROUP_M=8, USE_FAST_ACCUM=False, num_stages=5, num_warps=8, num_consumer_groups=0, num_buffers_warp_spec=0
SingleProcess AUTOTUNE benchmarking takes 3.6151 seconds and 0.1175 seconds precompiling for 28 choices
```

Reviewers: paulzhan, slarsen

Subscribers:

Tasks:

Tags:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
